Stratis.Bitcoin/BlockStore/BlockRepository.cs

```diff
@@ -293,40 +453,203 @@ public Task DeleteAsync(uint256 newlockHash, List<uint256> hashes)
 			Guard.NotNull(newlockHash, nameof(newlockHash));
 			Guard.NotNull(hashes, nameof(hashes));
 
-			return this.session.Do(() =>
-			{
-				foreach (var hash in hashes)
-				{
-					// if the block is already in store don't write it again
-					var key = hash.ToBytes();
+            return this.session.Do(() =>
+            {
+                foreach (var hash in hashes)
+                {
+                    // if the block is already in store don't write it again
+                    var key = hash.ToBytes();
 
-					if (this.TxIndex)
-					{
-						var block = this.session.Transaction.Select<byte[], Block>("Block", key);
+                    if (this.TxIndex)
+                    {
+                        var block = this.session.Transaction.Select<byte[], Block>("Block", key);
                         if (block.Exists)
                         {
                             this.PerformanceCounter.AddRepositoryHitCount(1);
 
+                            var removals = new Dictionary<uint160, HashSet<byte[]>>();
+
                             foreach (var transaction in block.Value.Transactions)
                             {
+                                var trxId = transaction.GetHash().ToBytes();
+
                                 this.PerformanceCounter.AddRepositoryDeleteCount(1);
-                                this.session.Transaction.RemoveKey<byte[]>("Transaction", transaction.GetHash().ToBytes());
+                                this.session.Transaction.RemoveKey<byte[]>("Transaction", trxId);
+
+                                // Remove transaction reference from indexed addresses
+                                foreach (var addrN in this.IndexedTransactionOutputs(transaction))
+                                {
+                                    var addr = (uint160)addrN[0];
+                                    var value = trxId.ToBytes().Concat(((uint)addrN[1]).ToBytes());
+
+                                    HashSet<byte[]> list;
+                                    if (!removals.TryGetValue(addr, out list))
```

Comment:
Thanks, I did not! I'm getting back into C# :D. I guess there is a slight trade-off between readability (being able to easily find the declaration) and being more concise.